File: Semestrul 4/MPP/lab03/cs/Repository/Impl/ParticipantRepositoryImpl.cs
using System;
using System.Data;
using System.Data.SQLite;
using cs.Domain;
using cs.Repository.Abstracts;
using cs.Repository.Interfaces;

namespace cs.Repository.Impl;

public class ParticipantRepositoryImpl : AbstractDBRepository<long, Participant>, IParticipantRepository
{
    public ParticipantRepositoryImpl() : base("Participants", "id")
    {
    }

    protected override string GetInsertParameters()
    {
        return "first_name, last_name, age, gender";
    }

    protected override string GetUpdateParameters()
    {
        return "first_name=@firstName, last_name=@lastName, age=@age, gender=@gender";
    }

    protected override int GetUpdateParameterCount()
    {
        return 4;
    }

    protected override Participant CreateEntity(IDataReader reader)
    {
        long id = Convert.ToInt64(reader["id"]);
        string firstName = reader["first_name"].ToString();
        string lastName = reader["last_name"].ToString();
        int age = Convert.ToInt32(reader["age"]);
        string gender = reader["gender"].ToString();

        Participant participant = new Participant(firstName, lastName, age, gender);
        participant.SetId(id);
        return participant;
    }

    protected override void SetParameters(SQLiteCommand command, Participant entity)
    {
        command.Parameters.AddWithValue("@firstName", entity.GetFirstName());
        command.Parameters.AddWithValue("@lastName", entity.GetLastName());
        command.Parameters.AddWithValue("@age", entity.GetAge());
        command.Parameters.AddWithValue("@gender", entity.GetGender());
    }

    public Participant FindByName(string name)
    {
        string sql = "SELECT * FROM Participants WHERE first_name || ' ' || last_name = @name";

        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            command.Parameters.AddWithValue("@name", name);

            using var reader = command.ExecuteReader();
            if (reader.Read())
            {
                return CreateEntity(reader);
            }
        }
        catch (Exception ex)
        {
            // Logger.LogError($"Error finding participant by name: {ex.Message}");
        }

        return null;
    }

    public Participant FindByEmail(string email)
    {
        string sql = "SELECT * FROM Participants WHERE email = @email";

        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            command.Parameters.AddWithValue("@email", email);

            using var reader = command.ExecuteReader();
            if (reader.Read())
            {
                return CreateEntity(reader);
            }
        }
        catch (Exception ex)
        {
            // Logger.LogError($"Error finding participant by email: {ex.Message}");
        }

        return null;
    }
}

File: Semestrul 4/SDI/lab14/java/TriathlonPersistence/src/main/java/triathlon/persistence/impl/ParticipantRepositoryImpl.java
package triathlon.persistence.impl;

import triathlon.model.Participant;
import triathlon.persistence.interfaces.*;
import triathlon.persistence.abstracts.AbstractDBRepository;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class ParticipantRepositoryImpl extends AbstractDBRepository<Long, Participant> implements ParticipantRepository {

    public ParticipantRepositoryImpl() {
        super("Participants", "id");
    }

    @Override
    protected String getInsertParameters() {
        return "first_name, last_name, age, gender";
    }

    @Override
    protected String getUpdateParameters() {
        return "first_name=?, last_name=?, age=?, gender=?";
    }

    @Override
    protected int getUpdateParameterCount() {
        return 4;
    }

    @Override
    protected Participant createEntity(ResultSet data) throws SQLException {
        Long id = data.getLong("id");
        String firstName = data.getString("first_name");
        String lastName = data.getString("last_name");
        int age = data.getInt("age");
        String gender = data.getString("gender");

        Participant participant = new Participant(firstName, lastName, age, gender);
        participant.setId(id);
        return participant;
    }

    @Override
    protected void saveEntity(PreparedStatement preparedStatement, Participant entity) throws SQLException {
        preparedStatement.setString(1, entity.getFirstName());
        preparedStatement.setString(2, entity.getLastName());
        preparedStatement.setInt(3, entity.getAge());
        preparedStatement.setString(4, entity.getGender());
    }

    @Override
    public Participant findByName(String name) {
        String sql = "SELECT * FROM Participants WHERE first_name || ' ' || last_name = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, name);

            try (ResultSet resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    return createEntity(resultSet);
                }
            }
        } catch (SQLException e) {
//            logger.error("Error finding participant by name: " + e.getMessage());
        }

        return null;
    }

    @Override
    public Participant findByEmail(String email) {
        String sql = "SELECT * FROM Participants WHERE email = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, email);

            try (ResultSet resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    return createEntity(resultSet);
                }
            }
        } catch (SQLException e) {
//            logger.error("Error finding participant by email: " + e.getMessage());
        }

        return null;
    }
}